src/providers/trusted_service/context/key_management.rs

```diff
@@ -0,0 +1,146 @@
+                id,
+                policy: Some(KeyPolicy {
+                    usage: key_attrs.policy.usage_flags.try_into()?,
+                    alg: key_attrs.policy.permitted_algorithms.try_into()?,
+                }),
+            }),
+        });
+        let mut proto_resp = Box::new(GenerateKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::GenerateKey,
+            self.rpc_caller,
+        )?;
+
+        Ok(0)
+    }
+
+    pub fn destroy_key(&self, id: u32) -> parsec_interface::requests::Result<()> {
+        if !self.check_key_exists(id)? {
+            return Err(ResponseStatus::PsaErrorDoesNotExist);
+        }
+        let proto_req = Box::new(OpenKeyIn { id });
+        let mut proto_resp = Box::new(OpenKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::OpenKey,
+            self.rpc_caller,
+        )?;
+        let handle = proto_resp.handle;
+
+        let proto_req = Box::new(DestroyKeyIn { handle });
+        let mut proto_resp = Box::new(DestroyKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::DestroyKey,
+            self.rpc_caller,
+        )?;
+        Ok(())
+    }
+
+    pub fn check_key_exists(&self, id: u32) -> Result<bool, Error> {
+        let proto_req = Box::new(OpenKeyIn { id });
+        let mut proto_resp = Box::new(OpenKeyOut::default());
+        match Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::OpenKey,
+            self.rpc_caller,
+        ) {
+            Ok(_) => {
+                let handle = proto_resp.handle;
+                let proto_req = Box::new(CloseKeyIn { handle });
+                Context::send_request(
+                    &proto_req,
+                    &mut proto_resp,
+                    Opcode::OpenKey,
+                    self.rpc_caller,
+                )?;
+                Ok(true)
+            }
+            Err(e) => {
+                if e == Error::DoesNotExist {
+                    Ok(false)
+                } else {
+                    Err(e)
+                }
+            }
+        }
+    }
+
+    fn send_request<Req: Message, Resp: Message>(
+        req: &Box<Req>,
+        resp: &mut Box<Resp>,
+        opcode: Opcode,
+        rpc_cl: *mut rpc_caller,
+    ) -> Result<(), Error> {
+        info!("Beginning call to Trusted Service");
+        let mut buf_out = null_mut();
+
+        let call_handle = unsafe { rpc_caller_begin(rpc_cl, &mut buf_out, req.encoded_len()) };
+        if call_handle == null_mut() {
+            error!("Call handle was null");
+            return Err(Error::CommunicationFailure);
+        }
+        let mut buf_out = unsafe { Vec::from_raw_parts(buf_out, 0, req.encoded_len()) };
+        req.encode(&mut buf_out).unwrap();
+        forget(buf_out);
```

Comment:
hmm I don't understand this part: why do we create a vec and then forget it 😕 ?

Reply:
Same kinda thing that was happening with the context pointers - creating a `Vec` from a pointer so that I can use it on the line above to pull data into a protobuf object, then forgetting it because it's the RPC caller who ultimately owns it and frees it in `rpc_caller_end`.

Reply:
Ah, so if I understand:

1. `rpc_caller_begin` allocates `req.encoded_len()` bytes at `buf_out`
2. we create a `Vec` over this location that we can pass to `req.encode`
3. `req.encode` writes `encoded_len()` bytes in the vector
4. We forget the `Vec` to prevent double-free

The RPC layer then expect the request to be there for the following operations and use it.

I think instead of creating a whole new `Vec` and forgetting it, we can just [create a slice from raw parts](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html), that way the memory management is not done by Rust and we can still (probably) pass it to `prost`.

Reply:
Nice, cheers for the link - had to use the `_mut` one but nonetheless, works fine.